File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int arr[] = new int[128];
        int l = 0, max = 0;
        Arrays.fill(arr, -1);
        for (int i=0; i<s.length(); i++) {
            int j = s.charAt(i);
            if (arr[j] >= l) { //aar[j] = -1 initially
                l = arr[j]+1;
            }
            arr[j] = i;
            max = Math.max(max, i-l+1);
        }
        return max;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ch = set()
        c, i, j = 0, 0, 0
        for j in range(len(s)):
            while s[j] in ch:
                ch.remove(s[i])
                i+=1
            ch.add(s[j])
            c = max(c, j-i+1)
        return c